libraries/node-core-library/src/JsonFile.ts

```diff
@@ -1,9 +1,107 @@
 // Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
 // See LICENSE in the project root for license information.
 
+import * as fsx from 'fs-extra';
+import * as os from 'os';
+import * as jju from 'jju';
+import Validator = require('z-schema');
+
+export type ValidateErrorCallback = (errorDescription: string) => void;
+
 /**
  * Utilities for reading/writing JSON files.
- * @alpha
+ * @public
  */
 export class JsonFile {
+
+  public static validateSchema(jsonObject: Object, jsonSchemaObject: Object,
+    errorCallback: ValidateErrorCallback): void {
+
+    // Remove the $schema reference that appears in the configuration object (used for IntelliSense),
+    // since we are replacing it with the precompiled version.  The validator.setRemoteReference()
+    // API is a better way to handle this, but we'd first need to publish the schema file
+    // to a public web server where Visual Studio can find it.
+    // tslint:disable-next-line:no-string-literal
+    delete jsonSchemaObject['$schema'];
+
+    const validator: Validator = new Validator({
+      breakOnFirstError: false,
+      noTypeless: true
+    });
+
+    if (!validator.validate(jsonObject, jsonSchemaObject)) {
+      const errorDetails: Validator.SchemaErrorDetail[] = validator.getLastErrors();
+
+      let buffer: string = 'JSON schema validation failed:';
+
+      buffer = JsonFile._formatErrorDetails(errorDetails, '  ', buffer);
+      errorCallback(buffer);
+    }
+  }
+
+  public static loadJsonFile(jsonFilename: string): {} {
+    if (!fsx.existsSync(jsonFilename)) {
+      throw new Error(`Input file not found: ${jsonFilename}`);
+    }
+
+    const buffer: Buffer = fsx.readFileSync(jsonFilename);
+    try {
+      return jju.parse(buffer.toString());
+    } catch (error) {
+      throw new Error(`Error reading "${jsonFilename}":` + os.EOL + `  ${error.message}`);
+    }
+  }
+
+  public static saveJsonFile(jsonFilename: string, jsonData: {}): void {
+    JsonFile._validateNoUndefinedMembers(jsonData);
+    const stringified: string = JSON.stringify(jsonData, undefined, 2) + '\n';
+    const normalized: string = JsonFile._getAllReplaced(stringified, '\n', '\r\n');
+    fsx.writeFileSync(jsonFilename, normalized);
+  }
+
+  /**
+   * Used to validate a data structure before writing.  Reports an error if there
+   * are any undefined members.
+   */
+  // tslint:disable-next-line:no-any
+  private static _validateNoUndefinedMembers(json: any): void {
+    if (!json) {
+      return;
+    }
+    if (typeof json === 'object') {
+      for (const key of Object.keys(json)) {
+        // tslint:disable-next-line:no-any
+        const value: any = json[key];
+        if (value === undefined) {
+          throw new Error(`The key "${key}" is undefined`);
+        }
+        JsonFile._validateNoUndefinedMembers(value);
+      }
+    }
+
+  }
+
+  private static _formatErrorDetails(errorDetails: Validator.SchemaErrorDetail[], indent: string,
```

Comment:
Get rid of all the other copies of Jsonfile

---
In reply to: [135151384](https://github.com/Microsoft/web-build-tools/pull/330#discussion_r135151384) [](ancestors = 135151384)